doc/how-to-design.md

```diff
@@ -0,0 +1,710 @@
+(e.g., with the offset+length of the syntactically-previous field, or the IR
+equivalent of `a < b && b < c`), while others require extensive changes to how
+code is generated.
+
+*   What information will the back end need in order to generate working code?
+*   Does this feature require embedded-unfriendly generated code?  (E.g.,
+    memory allocation, I/O.)
+*   Can the existing C++ back end, which just walks the IR tree in a single
+    pass while building up strings which are combined into a `.h`, handle this
+    feature in its current design?
+*   How will this feature interact with various generated templates?
+*   Can/should this feature be, itself, templated?
+
+
+## C++ Runtime Library
+
+The runtime library will be included with every program that touches Emboss, so
+it is important to make it efficient.  When adding features, consider:
+
+*   Can the feature be added in such a way that it does not cost anything for
+    programs that do not use the feature?  A standalone C++ template will not
+    be included in a program unless the program instantiates the template, but
+    if the new code is used from somewhere in an existing function, it may be
+    included in programs that do not use it directly.
+
+*   Can the feature be added without allocating any heap memory?  Can it be
+    added with O(1) stack memory use?  Both of these are important for some
+    embedded systems, such as OS-less microcontroller and hard-real-time
+    environments.  Some features may intrinsically require memory allocation,
+    in which case it is best if they can be separated: for example, Emboss
+    structure-to-string conversion requires allocation, and even `#include`'ing
+    the appropriate headers can be too much for some environments, even if the
+    serialization code is never included in the final binary.
+
+*   How much can you rely on the C++ compiler to optimize things?  If you have
+    to implement your own optimizations, that will cost more development time
+    and add more complexity to the standard library.
+
+
+## Compiler Complexity
+
+The Emboss compiler is already quite complex, and has many subsystems that
+interact.  It is already quite difficult to reason about some interactions.
+
+*   Can the feature be added at an "edge" of the compiler?  For example, if you
+    can implement your feature as syntax sugar that converts the new feature to
+    existing IR early in the compilation process, it is much easier to verify
+    that it will not cause problematic interactions.  Similarly, if you can
+    implement your feature entirely in the back end or in the runtime library,
+    you do not need to worry about interactions inside the front end.
+
+*   If a feature cannot be added at an edge, how can you design it to minimize
+    the complexity?  (Ideally, you could even unify existing systems in such a
+    way that the overall complexity of the compiler is lower at the end.)
+
+
+## Future Back Ends
```

Comment:
It might be useful to move this section to [`roadmap.md`](https://github.com/google/emboss/blob/master/doc/roadmap.md) and then link to the heading in this doc or a footnote, that way `roadmap.md` can be the SOT for this kind of information (`roadmap.md` already contains a couple of these potential backends listed).

The reason I'm suggesting this is this is something that could change over time as other backend use cases become relevant and the project evolves, and it would be easier to reference a single SOT rather than have these sorts of ideas scattered about potentially multiple docs.

Reply:
This is a reasonable suggestion, but `roadmap.md` is in pretty bad shape at the moment, so I'd like to defer that until I have a chance to update the roadmap.

Reply:
Sounds good, that works